### ISSUE TITLE
NuGet.exe on the agent should specify binding redirects for MSBuild

### ISSUE THREAD

**Issue by duncanpMS** [Microsoft/vso-agent-tasks#988]
As a result of #529 the user can now specify that `nuget pack` should trigger a build. The user can also specify which version of MSBuild should be used. However, *nuget.exe* doesn't specify binding redirects for MSBuild which can cause the build step to fail 
e.g. if the user specifies `-MSBuild 12` but the project being built uses a target built against MSBuild v4.0 the build will fail.
See #793 and #972 for more information.

Workaround: on a hosted agent, add the following content in a text file called *nuget.exe.config* next to *nuget.exe* i.e. in *\agent\Worker\Tools\*:

```xml
<?xml version ="1.0"?>
<configuration>
    <runtime>
      <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
        <dependentAssembly>
          <assemblyIdentity name="Microsoft.Build.Framework" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
          <bindingRedirect oldVersion="2.0.0.0-12.0.0.0" newVersion="14.0.0.0"/>
        </dependentAssembly>
        <dependentAssembly>
          <assemblyIdentity name="Microsoft.Build" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
          <bindingRedirect oldVersion="2.0.0.0-12.0.0.0" newVersion="14.0.0.0"/>
        </dependentAssembly>
        <dependentAssembly>
          <assemblyIdentity name="Microsoft.Build.Engine" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
          <bindingRedirect oldVersion="2.0.0.0-12.0.0.0" newVersion="14.0.0.0"/>
        </dependentAssembly>
        <dependentAssembly>
          <assemblyIdentity name="Microsoft.Build.Conversion.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
          <bindingRedirect oldVersion="2.0.0.0-12.0.0.0" newVersion="14.0.0.0"/>
        </dependentAssembly>
        <dependentAssembly>
          <assemblyIdentity name="Microsoft.Build.Tasks.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
          <bindingRedirect oldVersion="2.0.0.0-12.0.0.0" newVersion="14.0.0.0"/>
        </dependentAssembly>
        <dependentAssembly>
          <assemblyIdentity name="Microsoft.Build.Utilities.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
          <bindingRedirect oldVersion="2.0.0.0-12.0.0.0" newVersion="14.0.0.0"/>
        </dependentAssembly>
      </assemblyBinding>
    </runtime>
</configuration>
```

**Comment by vtbassmatt** [comment]
I'm not sure we can do anything correct here. We can redirect to 14.0.0.0 but then a user that needs 12.0.0.0 is stuck, or vice versa. Since we allow you to bring-your-own-nuget.exe in the NuGet tasks, you could also put nuget.exe.config in the repo. Am I missing anything?

**Comment by duncanpMS** [comment]
@pascalberger does the proposed work for you?
@vtbassmatt Users should be able to work round the issue if they check in their own nuget.exe and provide a suitable config file, but this solution isn't very discoverable - could it be made more so?
Also, I'd guess that the current behaviour will fail for most users that use the -MSBuild option i.e. it will fail unless the build only uses tasks from a single version of MSBuild. MSBuild assemblies are backwards compatible so providing a redirect to the latest version will allow builds that uses tasks built against different versions to succeed. That would seem like a better default behaviour for most users. If a user *really* needed to build using a specific down-level version of MSBuild for some reason then they could do that by checking in a nuget.exe with an appropriate nuget.exe.config.

**Comment by pascalberger** [comment]
@duncanpMS @vtbassmatt Our case is a little special that we have quite a few repositories. Therefore putting nuget.exe.config in each repository just for the build server is not really a solution since it is not easy to maintain. Packaging it in a custom build task might be a solution, but currently we're working around it by first building the source for sonar and in a second step package the stuff.

**Comment by vtbassmatt** [comment]
nuget.exe will accept `-MSBuildVersion` as an argument. Does this solve the problem more cleanly than checking in binding redirects to a bunch of repos?

**Comment by duncanpMS** [comment]
@vtbassmatt no, specifying *-MSBuildVersion* doesn't help - we tried that when looking at issues #793 and #972.
The underlying issue appears to be with how *NuGet.exe* triggers MSBuild; it looks for the appropriate version of MSBuild, then executes it using the MSBuild API rather than executing *MSBuild.exe*. *NuGet.exe* doesn't provide any binding redirects so the build failures will be common (i.e. unless every task used in the build is built against the same version of MSBuild). If *NuGet.exe* executed *MSBuild.exe* then the issue probably wouldn't arise as each version of *MSBuild.exe* does provide binding redirects for earlier versions: see e.g.
%PROGRAMFILES(x86)%\MSBuild\12.0\bin\msbuild.exe.config and 
%PROGRAMFILES(x86)%\MSBuild\14.0\bin\msbuild.exe.config

However, if the task shipped with a *NuGet.exe.config* that had the same redirects as *%PROGRAMFILES(x86)%\MSBuild\14.0\bin\msbuild.exe.config* then any build triggered by the task would be much more likely to succeed.

**Comment by jmyersmsft** [comment]
if we add a redirect to v14, and the user specifies -MSBuildVersion 12, does that not mean that we are either a) broken, or b) actually using 14 when the user explicitly wants 12?

**Comment by duncanpMS** [comment]
a) probably not since the assemblies are backwards-compatible, and b) this may or may not be a problem. The v14 backwards-compatible assemblies would be used which should be ok. The important thing is whether the v12 targets and tasks are used, as they control how the build is actually performed.
